akka-javasdk/src/main/scala/akka/javasdk/impl/eventsourcedentity/EventSourcedEntityImpl.scala

```diff
@@ -0,0 +1,292 @@
+          }
+
+        case NoPrimaryEffect =>
+          val (reply, error) = replyOrError(updatedState)
+
+          Future.successful(
+            new SpiEventSourcedEntity.Effect(events = Vector.empty, updatedState = state, reply, error, None))
+      }
+
+    } catch {
+      case CommandHandlerNotFound(name) =>
+        throw new EntityExceptions.EntityException(
+          entityId,
+          0, // FIXME remove commandId
+          command.name,
+          s"No command handler found for command [$name] on ${router.entity.getClass}")
+      case BadRequestException(msg) =>
+        Future.successful(
+          new SpiEventSourcedEntity.Effect(
+            events = Vector.empty,
+            updatedState = state,
+            reply = None,
+            error = Some(new SpiEntity.Error(msg, Status.Code.INVALID_ARGUMENT.value)),
+            delete = None))
+      case e: EntityException =>
+        throw e
+      case NonFatal(error) =>
+        throw EntityException(
+          entityId = entityId,
+          commandId = 0,
+          commandName = command.name,
+          s"Unexpected failure: $error",
+          Some(error))
+    } finally {
+      router.entity._internalSetCommandContext(Optional.empty())
+      router.entity._internalClearCurrentState()
+      cmdContext.deactivate() // Very important!
+
+      span.foreach { s =>
+        MDC.remove(Telemetry.TRACE_ID)
+        s.end()
+      }
+    }
+
+  }
+
+  override def handleEvent(
+      state: SpiEventSourcedEntity.State,
+      eventEnv: SpiEventSourcedEntity.EventEnvelope): SpiEventSourcedEntity.State = {
+    val event =
+      messageCodec
+        .decodeMessage(eventEnv.payload)
+        .asInstanceOf[AnyRef] // FIXME empty?
+    entityHandleEvent(state, event, eventEnv.entityId, eventEnv.sequenceNumber)
+  }
+
+  def entityHandleEvent(
+      state: SpiEventSourcedEntity.State,
+      event: AnyRef,
+      entityId: String,
+      sequenceNumber: Long): SpiEventSourcedEntity.State = {
+    val eventContext = new EventContextImpl(entityId, sequenceNumber)
+    val router = createRouter(eventContext) // FIXME reuse router instance?
+    router.entity._internalSetEventContext(Optional.of(eventContext))
+    try {
+      router.handleEvent(state, event)
+    } catch {
+      case EventHandlerNotFound(eventClass) =>
+        throw new IllegalArgumentException(s"Unknown event type [$eventClass] on ${router.entity.getClass}")
+    } finally {
+      router.entity._internalSetEventContext(Optional.empty())
+    }
+  }
+
+  override val stateSerializer: SpiSerialization.Serializer =
+    new SpiSerialization.Serializer {
+
+      override def toProto(obj: Deserialized): ScalaPbAny =
+        ScalaPbAny.fromJavaProto(messageCodec.encodeJava(obj))
+
+      override def fromProto(pb: ScalaPbAny): Deserialized =
+        messageCodec.decodeMessage(pb).asInstanceOf[Deserialized]
```

Comment:
this is incorrect, I'll follow up, see https://github.com/akka/akka-sdk/pull/39#discussion_r1862463386